includes/Modules/Tag_Manager/Settings.php

```diff
@@ -0,0 +1,86 @@
+<?php
+/**
+ * Class Google\Site_Kit\Modules\Tag_Manager\Settings
+ *
+ * @package   Google\Site_Kit\Modules\Tag_Manager
+ * @copyright 2019 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Modules\Tag_Manager;
+
+use Google\Site_Kit\Core\Storage\Setting;
+use Google\Site_Kit\Core\Util\Migrate_Legacy_Keys;
+
+/**
+ * Class for Tag Manager settings.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class Settings extends Setting {
+	use Migrate_Legacy_Keys;
+
+	const OPTION = 'googlesitekit_tagmanager_settings';
+
+	/**
+	 * Mapping of legacy keys to current key.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	protected $legacy_key_map = array(
+		'account_id'   => 'accountID',
+		'accountId'    => 'accountID',
+		'container_id' => 'containerID',
+		'containerId'  => 'containerID',
+	);
+
+	/**
+	 * Registers the setting in WordPress.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function register() {
+		register_setting(
+			self::OPTION,
+			self::OPTION,
+			array(
+				'type'              => 'object',
+				'sanitize_callback' => $this->get_sanitize_callback(),
+				'default'           => $this->get_default(),
+			)
+		);
+
+		add_filter(
+			'option_' . self::OPTION,
+			function ( $option ) {
+				if ( ! is_array( $option ) ) {
+					$option = $this->get_default();
+				} else {
+					$option = $this->migrate_legacy_keys( $option, $this->legacy_key_map );
+				}
+
+				// Fill in any missing keys with defaults.
+				return $option + $this->get_default();
+			}
+		);
```

Comment:
This filter is actually the same code in 3 of the 4 setting implementations as well, with `Analytics\Settings` including it too, just doing a bit more.

This should definitely not be in the base class because it is too specific, but maybe `Migrate_Legacy_Keys` could also include a `add_migrate_legacy_keys_option_filter()` (not this name), or something a little less specific, but still reusable :) Or there could be a `Setting_With_Legacy_Keys_Trait`.

Even the `Analytics\Settings` class could use this, it could then just add another `option_{$option}` filter for its additional logic. It shouldn't be problematic for these to be two individual filters.

Reply:
I opted for the trait, but also added a base `Module_Settings` class which is where the default handling is now as well as the new `get_type` since these will always be `object`.